zeebe/dynamic-config/src/main/java/io/camunda/zeebe/dynamic/config/api/ClusterPatchRequestTransformer.java

```diff
@@ -0,0 +1,92 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public final class ClusterPatchRequestTransformer implements ConfigurationChangeRequest {
+
+  private final ClusterPatchRequest request;
+  private final MemberId localMemberId;
+
+  public ClusterPatchRequestTransformer(
+      final ClusterPatchRequest request, final MemberId localMemberId) {
+    this.request = request;
+    this.localMemberId = localMemberId;
+  }
+
+  @Override
+  public Either<Exception, List<ClusterConfigurationChangeOperation>> operations(
+      final ClusterConfiguration clusterConfiguration) {
+    final var requestIsValid = validate(request, clusterConfiguration);
+
+    return requestIsValid.flatMap(ignore -> generateOperations(clusterConfiguration));
+  }
+
+  @Override
+  public boolean isForced() {
+    return request.force();
+  }
+
+  private Either<Exception, List<ClusterConfigurationChangeOperation>> generateOperations(
+      final ClusterConfiguration clusterConfiguration) {
+    if (request.force()) {
+      final var membersToRetain = new HashSet<>(clusterConfiguration.members().keySet());
+      membersToRetain.removeAll(request.membersToRemove());
+
+      return new ForceScaleDownRequestTransformer(membersToRetain, localMemberId)
+          .operations(clusterConfiguration);
+    }
+
+    if (request.newClusterSize().isPresent()) {
+      final var newSetOfMembers =
+          IntStream.range(0, request.newClusterSize().get())
+              .mapToObj(i -> MemberId.from(String.valueOf(i)))
+              .collect(Collectors.toSet());
+      return new ScaleRequestTransformer(
+              newSetOfMembers, request.newReplicationFactor(), request.newPartitionCount())
+          .operations(clusterConfiguration);
+    }
+
+    final var newSetOfMembers = new HashSet<>(clusterConfiguration.members().keySet());
+    newSetOfMembers.addAll(request.membersToAdd());
+    newSetOfMembers.removeAll(request.membersToRemove());
+
+    return new ScaleRequestTransformer(
+            newSetOfMembers, request.newReplicationFactor(), request.newPartitionCount())
+        .operations(clusterConfiguration);
+  }
+
+  private Either<Exception, Boolean> validate(
+      final ClusterPatchRequest request, final ClusterConfiguration configuration) {
+    if (request.newClusterSize().isPresent()
+        && (!request.membersToAdd().isEmpty() || !request.membersToRemove().isEmpty())) {
+      return Either.left(
+          new ClusterConfigurationRequestFailedException.InvalidRequest(
+              "Provide only one of 'newClusterSize' or 'membersToAdd'/'membersToRemove'"));
+    }
+
+    if (request.force()
+        && (request.newClusterSize().isPresent() || !request.membersToAdd().isEmpty())) {
+      return Either.left(
+          new ClusterConfigurationRequestFailedException.InvalidRequest(
+              "Force operation is only allowed when removing members"));
+    }
```

Comment:
:wrench: Similarly, changing the replication factor or partition count is also not supported.